pkg/command/service/load.go

```diff
@@ -0,0 +1,423 @@
+					}
+					fmt.Printf("\n[Verbose] Pods changed from 0 to %d:\n", len(loadResult.PodResults))
+					fmt.Printf("pods\tready_duration(seconds)\n")
+					for i := 0; i < len(loadResult.PodResults); i++ {
+						fmt.Printf("%9d\t%23.1f\n", i, loadResult.PodResults[i].PodReadyDuration)
+					}
+					fmt.Printf("\n---------------------------------------------------------------------------------\n")
+				}
+				m.Lock()
+				result.Measurment = append(result.Measurment, loadResult)
+				m.Unlock()
+			} else {
+				fmt.Printf("result of load is error: %s", err)
+			}
+		}(i, &m)
+	}
+	wg.Wait()
+
+	return result, nil
+}
+
+func runLoadFromZero(ctx context.Context, params *pkg.PerfParams, inputs pkg.LoadArgs, namespace string, svc *servingv1.Service) (
+	string, pkg.LoadFromZeroResult, error) {
+	selector := labels.SelectorFromSet(labels.Set{
+		serving.ServiceLabelKey: svc.Name,
+	})
+	var loadOutput string
+	var loadResult pkg.LoadFromZeroResult
+	var replicaResults []pkg.LoadReplicaResult
+	var podResults []pkg.LoadPodResult
+	loadResult.ServiceName = svc.Name
+	loadResult.ServiceNamespace = namespace
+
+	watcher, err := params.ClientSet.AppsV1().Deployments(namespace).Watch(
+		context.Background(), metav1.ListOptions{LabelSelector: selector.String()})
+	if err != nil {
+		m := fmt.Sprintf("unable to watch the deployment for the service: %v", err)
+		log.Println(m)
+		return "", loadResult, errors.New(m)
+	}
+	defer watcher.Stop()
+
+	rdch := watcher.ResultChan() // replica duration channel
+	pdch := make(chan struct{})  // pod duration channel
+	errch := make(chan error)
+
+	endpoint, err := resolveEndpoint(ctx, params, inputs.ResolvableDomain, svc)
+	if err != nil {
+		return "", loadResult, fmt.Errorf("failed to get the cluster endpoint: %w", err)
+	}
+
+	// Prepare for load test
+	cmd, wrkLua, err := loadCmdBuilder(inputs, endpoint, namespace, svc)
+	if err != nil {
+		return "", loadResult, err
+	}
+
+	defer func() {
+		// Delete wrk lua script
+		if strings.EqualFold(inputs.LoadTool, "wrk") {
+			_, fileError := os.Stat(wrkLua)
+			if fileError == nil {
+				removeError := os.Remove(wrkLua)
+				if removeError != nil {
+					log.Printf("remove %s error : %s", wrkLua, removeError)
+				}
+			}
+		}
+	}()
+
+	loadStart := time.Now()
+	m := fmt.Sprintf("Namespace %s, Service %s, load start", namespace, svc.Name)
+	log.Println(m)
+
+	go func() {
+		runCmd := exec.Command("/bin/sh", "-c", cmd)
+		var output []byte
+		output, err = runCmd.Output()
+		loadOutput = string(output)
+
+		if err != nil {
+			m := fmt.Sprintf("run load command error: %s", err)
+			fmt.Println(m)
+			return
+		}
+
+		loadEnd := time.Now()
+		loadDuration := loadEnd.Sub(loadStart)
+		m := fmt.Sprintf("Namespace %s, Service %s, load end, take off %.3f seconds", namespace, svc.Name, loadDuration.Seconds())
+		log.Println(m)
+
+		time.Sleep(inputs.WaitPodsReadyDuration * time.Second) //wait for all pods ready
+		pdch <- struct{}{}
+	}()
+
+	var preReadyReplicas int
+	for {
```

Comment:
Since this starts a key part of the code with "complicated" infinite loop with case/switch statement I am wondering if it could be simplified a little by perhaps converting the statements into the each case as a call to a helper function? Naming functions and refactoring to use these helpers could also help reveal commonality and perhaps simplify further. Just a thought

Reply:
Thank you, do you means the loop like this:
```go
for {
      select {
                case event := <-rdch:
	                function_1()
                case <-pdch:
	                function_2()
                case err := <-errch:
	                return nil, nil, err
      }
```